### ISSUE TITLE
Component stops listening the Store after first router transition

### ISSUE THREAD

**Issue by alexfedoseev** [acdlite/flummox#195]
I've just started using `flummox` and I'm a little bit flummoxed :)

Here is my use case. 

**App Layout**
```javascript
<section id="layout">
  <Header {...this.props} />
  <RouteHandler {...this.props} />
  <Footer />
  <Alert {...this.props} />
</section>
```
In my App I have `Alert Component`. When something happens I trigger an `AlertAction` from some component, it dispatches alert payload to `AlertStore`, which gets updated, and `AlertComponent` shows alert ( + hides it after some time).

For example I have a `PostEdit Component`. After form submit, I send request to API from `PostActions` and receive response, which is dispatched to `PostStore`. Store gets updated and `PostEdit Component` gets notified. In `PostEdit`'s `componentWillReceiveProps` I check props, received from the store, and trigger `AlertAction` to show the alert.

2 problems:

1. I have to use `setTimeout` to trigger `AlertAction` from the `Post Component` to make alert things happen (code below).
2. And the main problem is that `Alert Component` stops listening `AlertStore` after the first `react-router` transition.

Here is console.log, illustrating the problem:

![image](https://cloud.githubusercontent.com/assets/4244251/7767479/7b5d0c9a-007b-11e5-8d38-22f9558cabb9.png)

One more strange thing is that changed-store-notification in console.log printed before dispatch-payload-from-action-notification (which causes this store change).

Here are code snippets:

**AlertHandler.jsx**
```javascript
export default class AlertHandler extends React.Component {

  // constructor()

  render() {
    return (
        <FluxComponent connectToStores={'alerts'}>
          <Alert {...this.props} />
        </FluxComponent>
    );
  }
}
```

**Alert.jsx**
```javascript
export default class Alert extends React.Component {

  // constructor()
  // _handleShow()
  // _handleHide()

  componentDidMount() {
    this.props.flux.getStore('alerts').addListener('change', function() {
      console.log('Changed!', this.state);
    });
  }

  componentWillUnmount() {
    console.log('Not gonna happen');
  }

  // render()
}
```

**PostEdit.jsx**
```javascript
export default class PostEdit extends React.Component {

  // constructor()

  componentWillReceiveProps(newProps) {
    this.setState({
      isLoading: false
    }, () => {
      if (newProps.errors) {
        // without `setTimeout` nothing happens
        setTimeout(() => {
          newProps.flux
              .getActions('alerts')
              .showErrorAlert(newProps.errors);
        }, 1);
      } else {
        setTimeout(() => {
          newProps.flux
              .getActions('alerts')
              .showSuccessAlert();
        }, 1);
      }
    });
  }

  _submitPost(e) {

    // doing stuff...

    // triggering updatePost action
    this.props.flux
        .getActions('posts')
        .updatePost(post);
  }

  // render()
}
```

Not sure are these bugs or I missed smth in flux/flummox patterns and do things wrong. Thanks for feedback!

**Comment by johanneslumpe** [comment]
You are already rendering your `Alert` within a `FluxComponent`. You don't have to add another listener within the component itself. The store will automatically pass new props to your component whenever the `alerts` store emits a change.

I would move the action triggers out of `componentWillReceiveProps` and rather have your alert store listen to all relevant actions.

**Comment by alexfedoseev** [comment]
@johanneslumpe This listener is for `console.log` only, to make sure that store state has changed.

**Comment by alexfedoseev** [comment]
> I would move the action triggers out of componentWillReceiveProps and rather have your alert store listen to all relevant actions.

Can you post an example please?

**Comment by johanneslumpe** [comment]
Yeah sure, here you go:
```javascript
class AlertStore extends Store {
  
  constructor() {
    
    super();
    
    this. registerAllAsync(
      this.beginAllHandler,
      this.successAllHandler,
      this.failureAllHandler
    );
    
  }
  
  beginAllHandler(action) {
    // every single async begin action comes here
    // 
    // `action` looks like this:
    // {
    //   "actionId": "1-someaction",
    //   "dispatchId": 123,
    //   "actionArgs": [1,2,"arg"],
    //   "async":"begin"
    // }
  }
  
  successAllHandler(response, action) {
    // every single async success action comes here
    // 
    // `response` is the result from your async action
    // // `action` looks like this:
    // {
    //   "actionId": "1-someaction",
    //   "dispatchId": 123,
    //   "actionArgs": [1,2,"arg"],
    //   "async":"success",
    //   "body": {} // just an example, this will contain the response you
    //                 already get passed in as first parameter
    // }
  }
  
  failureAllHandler(err, action) {
    // every single async failure action comes here
    // 
    // just like the above, except that `action.async` will be "failure"
    // and that you have `action.error` instead of `action.body`
  }
  
}
```

Now you will be able to react to each async action with this store. Does this answer you question?

**Comment by alexfedoseev** [comment]
Thanks for example! I moved action triggers to Store, but the problem still there.

I created sample app to illustrate this issue and another one ( #219 ).
Sample app: https://github.com/alexfedoseev/flummox-cuts (clone repo && `npm start`)

It has list of posts. To reproduce issue click the button `Add new post`, it will create a post and fire an alert. After this click on some tag and navigate to filtered list of posts. Click the button again and there will be no alert.

P.S. And btw in `successAllHandler(response, action) {}` I have only one arg — `response`, `action` is `undefined`.
